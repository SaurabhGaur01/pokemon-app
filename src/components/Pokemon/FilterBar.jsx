import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import filterPokemons from '../../actions/filterPokemons';
import sortPokemons from '../../actions/sortPokemons';
import AppDropdown from '../shared/AppDropdown';
import { SORT_TYPES } from '../../constants/appConstants';

const FilterBar = ({
  filterPokemons,
  sortPokemons,
}) => {
  const inputOnchage = (event) => {
    filterPokemons(event.target.value);
  }

  const onSortChange = (event) => {
    sortPokemons(event.target.value);
  }

  return (
    <div className="row">
      <div className="center-child col-md-4">
        <AppDropdown
          label="Sort By"
          options={SORT_TYPES}
          onSelection={onSortChange}
        />
     </div>

      <div className="center-child col-md-4">
        <div className="input-group input-group-sm mb-3">
            <div className="input-group-prepend">
                <span className="input-group-text" id="inputGroup-sizing-sm">
                    Search By
                </span>
            </div>
            <input
              id="name-input"
              type="text"
              className="form-control"
              aria-label="Small"
              aria-describedby="inputGroup-sizing-sm"
              onChange={inputOnchage}
              placeholder="Name or Abilities" 
            />
        </div>
      </div>
  </div>
  )
}

FilterBar.propTypes = {
  filterPokemons: PropTypes.func.isRequired,
};

const mapDispatchToProps = {
  filterPokemons,
  sortPokemons,
};

export { FilterBar as TestableFilterBar };

export default connect(null, mapDispatchToProps)(FilterBar);