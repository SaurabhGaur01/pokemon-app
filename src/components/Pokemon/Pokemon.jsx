import React from 'react';
import PropTypes from 'prop-types';
import { withRouter, Link } from 'react-router-dom';

const Pokemon = ({ pokemon }) => {
  const {
    id, name, sprites, height, weight, abilities
  } = pokemon;
  const abilitiesList = abilities.map(ability => ability.ability.name);
  return (
    <div>
      <Link className="pokemon-card" to={`/pokemon/${id}`} id="link-detail">
        <div className="d-flex-around title-card">
          <span>
            #
            {id}
          </span>
          <span data-testid="pokemon-name">
            {name && name.charAt(0).toUpperCase() + name.slice(1)}
          </span>
        </div>
        <img className="card-image" src={sprites.front_default} alt={name} />
        <span className="type-title">Height - {height}</span>
        <span className="type-title">Weight - {weight}</span>
        <div className="type-container">
          <span className="type-title">Abilities</span>
          <ul className="type-list">
            {
              abilitiesList && abilitiesList.map(ability => (<li key={ability}>{ability}</li>))
            }
          </ul>
        </div>
      </Link>
    </div>
  );
};

Pokemon.propTypes = {
  pokemon: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    sprites: PropTypes.shape({}).isRequired,
    types: PropTypes.arrayOf.isRequired,
    height: PropTypes.number.isRequired,
    weight: PropTypes.number.isRequired,
    abilities: PropTypes.arrayOf(PropTypes.shape({})),
  }).isRequired,
};

export default withRouter(Pokemon);