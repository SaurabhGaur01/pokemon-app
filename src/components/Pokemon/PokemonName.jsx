import React, { Component } from 'react';
import PropTypes from 'prop-types';

class PokemonName extends Component {
  constructor(props) {
    super(props);
    this.name = React.createRef();
  }

  render() {
    const { changeName, filterName } = this.props;
    return (
      <div className="margin-y">
        <label htmlFor="txtName">
          Search By:
          <input data-testid="filter-name" className="margin-left" type="text" id="txtName" name="txtName" value={filterName} onChange={() => changeName(this.name.value.toLowerCase())} ref={input => { (this.name = input); }} placeholder="Name or Abilities" />
        </label>
      </div>
    );
  }
}

PokemonName.propTypes = {
  changeName: PropTypes.func.isRequired,
  filterName: PropTypes.string.isRequired,
};

export default PokemonName;