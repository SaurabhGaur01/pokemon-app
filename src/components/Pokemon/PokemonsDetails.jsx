import React from 'react';
import { connect } from 'react-redux';
import { useParams } from 'react-router-dom';
import PropTypes from 'prop-types';
import loadImg from '../../assets/loadImg.gif';
import handleRetrievePokemonDetails from '../../actions/handleRetrievePokemonDetails';

const PokemonsDetails = ({
    handleRetrievePokemonDetails,
    pokemonDetails,
    loading,
}) => {
    const { number } = useParams();

    React.useEffect(() => {
        handleRetrievePokemonDetails(number)
    }, []);

    return (
        <React.Fragment>
            { loading
                && (
                <div data-testid="loading-det" className="bg-load">
                    <img className="margin-top-30 image-load" src={loadImg} alt="loadingImage" />
                </div>
            )}

            { !loading && (
                <div className="pok-details" key={`#${pokemonDetails.number} ${pokemonDetails.namePkm}`}>
                    <div className="main-info-container">
                        <span className="detail-name">
                        {`#${pokemonDetails.number} ${pokemonDetails.namePkm}`.toUpperCase()}
                        </span>
                        <div className="flex-details card-detail-container">
                        <div className="detail-col-1">
                            <img className="detail-img" src={pokemonDetails.image} alt={pokemonDetails.namePkm} />
                            <div className="detail-text">
                            {pokemonDetails.text}
                            </div>
                        </div>
                        <div className="flex-details-2">
                            <div className="mr-2">
                                <h4>Details Pokemon</h4>
                                <div data-testid="pokemon-height" className="my-3">
                                    <span className="font-weight-bold">Height: </span>
                                    {pokemonDetails.height}
                                </div>
                                <div data-testid="pokemon-weight" className="my-3">
                                    <span className="font-weight-bold">Weight: </span>
                                    {pokemonDetails.weight}
                                </div>
                                <div className="my-3">
                                    <span className="font-weight-bold">Capture Rate: </span>
                                    {pokemonDetails.captureRate}
                                </div>
                                <div className="my-3">
                                    <span className="font-weight-bold">Growth Rate: </span>
                                    {pokemonDetails.growthRate}
                                </div>
                                <div className="my-3">
                                    <span className="font-weight-bold">Base Exp: </span>
                                    {pokemonDetails.baseExperience}
                                </div>
                                <div className="type-container">
                                    { pokemonDetails.types && pokemonDetails.types.length === 1
                                    && (
                                    <span className="type-title">Type</span>
                                    )}
                                    { pokemonDetails.types && pokemonDetails.types.length > 1
                                    && (
                                        <span className="type-title">Types</span>
                                    )}
                                    <ul className="type-list-det">
                                    {
                                        pokemonDetails.types
                                        && pokemonDetails.types.map(type => (<li key={type}>{type}</li>))
                                    }
                                    </ul>
                                </div>
                            </div>
                            <div>
                                <h4>Details Species</h4>
                                <div className="my-3">
                                    <span className="font-weight-bold">Habitat: </span>
                                    { pokemonDetails.habitat && pokemonDetails.habitat.charAt(0).toUpperCase() + pokemonDetails.habitat.slice(1) }
                                </div>
                                <div className="my-3">
                                    <span className="font-weight-bold">Shape: </span>
                                    { pokemonDetails.shape && pokemonDetails.shape.charAt(0).toUpperCase() + pokemonDetails.shape.slice(1) }
                                </div>
                                <div className="my-3">
                                    <span className="font-weight-bold">Color: </span>
                                    { pokemonDetails.color && pokemonDetails.color.charAt(0).toUpperCase() + pokemonDetails.color.slice(1) }
                                </div>
                            </div>
                        </div>
                        </div>
                    </div>
                </div>
            )}
        </React.Fragment>
    )
}

PokemonsDetails.propTypes = {
    pokemonDetails: PropTypes.shape({}),
    loading: PropTypes.bool.isRequired,
    handleRetrievePokemonDetails: PropTypes.func.isRequired,
}

const mapDispatchToProps = {
    handleRetrievePokemonDetails,
};

const mapStateToProps = state => ({
    pokemonDetails: state.pokemonDetails,
    loading: state.loading,
});

export { PokemonsDetails as TestablePokemonsDetails };

export default connect(mapStateToProps, mapDispatchToProps)(PokemonsDetails);