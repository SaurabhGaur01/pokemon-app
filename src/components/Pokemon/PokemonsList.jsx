import React, { useState } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import usePagination from "../shared/Pagination";
import Pokemon from './Pokemon';
import FilterBar from './FilterBar';
import loadImg from '../../assets/loadImg.gif';
import AppPaging from '../shared/AppPaging';

const PokemonsList = ({
    pokemons: { viewList }, loading,
}) => {
    let [page, setPage] = useState(1);
    const [rowsPerPage, setRowsPerPage] = React.useState(10);
    const count = Math.ceil(viewList.length / rowsPerPage);
    const filteredData = usePagination(viewList, rowsPerPage);
    
    const handleChange = (e, p) => {
        setPage(p);
        filteredData.jump(p);
    };
  
    const handleChangeRowsPerPage = (value) => {
        setRowsPerPage(parseInt(value, 10));
        setPage(0);
    };
    
    return (
        <React.Fragment>
            <div className="filter-container">
                <FilterBar />
                <AppPaging 
                    page={page} 
                    count={count} 
                    handleChange={handleChange}
                    handleChangeRowsPerPage={handleChangeRowsPerPage} 
                />
            </div>

            { loading && (
                <div data-testid="loading" className="bg-load">
                    <img className="image-load" src={loadImg} alt="loadingImage" />
                </div>
            )}

            <div className="cards-container">
                {
                    !loading && filteredData.currentData().map((pokemon, index) => (
                        <Pokemon key={index} pokemon={pokemon} />
                    ))
                }
            </div>

            <AppPaging 
                page={page} 
                count={count} 
                handleChange={handleChange}
                handleChangeRowsPerPage={handleChangeRowsPerPage} 
            />
        </React.Fragment>
    )
}

PokemonsList.propTypes = {
    loading: PropTypes.bool.isRequired,
    pokemons: PropTypes.shape({
        viewList: PropTypes.arrayOf(PropTypes.shape({})),
    }),
};

const mapStateToProps = state => ({
    pokemons: state.pokemons,
    loading: state.loading,
});
  
export default connect(mapStateToProps)(PokemonsList);