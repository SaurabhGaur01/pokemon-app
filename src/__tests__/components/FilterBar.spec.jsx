import React from 'react';
import { shallow } from 'enzyme';
import { TestableFilterBar as FilterBar } from '../../components/Pokemon/FilterBar';

let mockProps;

describe('FilterBar', () => {
    beforeEach(() => {
        mockProps = {
            filterPokemons: jest.fn(),
            sortPokemons: jest.fn(),
        }
    })
    it('should render component correctly', () => {
        const renderedModule = shallow(<FilterBar {...mockProps} />);
        expect(renderedModule).toMatchSnapshot();
    });

    it('should call filterPokemons name or abilities is searched', () => {
        const renderedModule = shallow(<FilterBar {...mockProps} />);
        renderedModule.find('#name-input').at(0).simulate('change', {target: { value: 'mockValue'}});
        expect(mockProps.filterPokemons).toHaveBeenCalledTimes(1);
    });
})