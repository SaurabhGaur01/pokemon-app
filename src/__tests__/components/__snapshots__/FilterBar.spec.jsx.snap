// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterBar should render component correctly 1`] = `
<div
  className="row"
>
  <div
    className="center-child col-md-4"
  >
    <AppDropdown
      label="Sort By"
      onSelection={[Function]}
      options={
        Array [
          Object {
            "text": "Please Select",
            "value": "NONE",
          },
          Object {
            "text": "Name ASC",
            "value": "NAME_ASC_ORDER",
          },
          Object {
            "text": "Name DSC",
            "value": "NAME_DSC_ORDER",
          },
          Object {
            "text": "Height ASC",
            "value": "HEIGHT_ASC_ORDER",
          },
          Object {
            "text": "Height DSC",
            "value": "HEIGHT_DSC_ORDER",
          },
          Object {
            "text": "Width ASC",
            "value": "WEIGHT_ASC_ORDER",
          },
          Object {
            "text": "Width DSC",
            "value": "WEIGHT_DSC_ORDER",
          },
        ]
      }
    />
  </div>
  <div
    className="center-child col-md-4"
  >
    <div
      className="input-group input-group-sm mb-3"
    >
      <div
        className="input-group-prepend"
      >
        <span
          className="input-group-text"
          id="inputGroup-sizing-sm"
        >
          Search By
        </span>
      </div>
      <input
        aria-describedby="inputGroup-sizing-sm"
        aria-label="Small"
        className="form-control"
        id="name-input"
        onChange={[Function]}
        placeholder="Name or Abilities"
        type="text"
      />
    </div>
  </div>
</div>
`;
